### PR TITLE
fix(envoy): Use pod service names for rds checks

### PR DESCRIPTION
**Pull request by johnsonshi**
Check pod service names when performing rds route domain checks.

Currently, the pre-existing rds route domain checks only check for
the existence of the pod name in RDS routes. The pre-existing
check's logic is incorrect and this causes the RDS checks to
always fail.

The correct way to check for RDS routes is to check pod service
names, not pod names.

Signed-off-by: Johnson Shi <Johnson.Shi@microsoft.com>

### CODE REVIEW

**Comment by johnsonshi** [comment]
Need to check for pod services in envoy dynamic route configs instead of checking for pod names in envoy dynamic route configs.

**Comment by johnsonshi** [comment]
Need to check for pod services in envoy dynamic route configs instead of checking for pod names in rds envoy configs.

image

File: pkg/connectivity/pod-to-pod.go [hunk 1]
```diff
@@ -96,10 +96,10 @@ func PodToPod(srcPod *corev1.Pod, dstPod *corev1.Pod, osmControlPlaneNamespace s
 		envoy.NewSpecificEndpointCheck(srcConfigGetter, dstPod),
 
 		// Check whether the source Pod has an outbound dynamic route config domain that matches the destination Pod.
-		envoy.NewOutboundRouteDomainPodCheck(srcConfigGetter, dstPod),
+		envoy.HasOutboundDynamicRouteConfigDomainPodCheck(client, srcConfigGetter, dstPod),
```
Comment (SanyaKochhar) [inline pkg/connectivity/pod-to-pod.go:99]:
```suggestion
		envoy.NewOutboundRouteDomainPodCheck(client, srcConfigGetter, dstPod),
```

File: pkg/connectivity/pod-to-pod.go [hunk 2]
```diff
@@ -96,10 +96,10 @@ func PodToPod(srcPod *corev1.Pod, dstPod *corev1.Pod, osmControlPlaneNamespace s
...
 
 		// Check whether the destination Pod has an inbound dynamic route config domain that matches the source Pod.
-		envoy.NewInboundRouteDomainPodCheck(dstConfigGetter, srcPod),
+		envoy.HasInboundDynamicRouteConfigDomainPodCheck(client, dstConfigGetter, srcPod),
```
Comment (SanyaKochhar) [inline pkg/connectivity/pod-to-pod.go:102]:
```suggestion
		envoy.NewInboundRouteDomainPodCheck(client, dstConfigGetter, srcPod),
```

File: pkg/connectivity/pod-to-pod.go (hunk 1 above)
Comment (SanyaKochhar) [reply pkg/connectivity/pod-to-pod.go:99]: ref: https://github.com/openservicemesh/osm-health/pull/74#discussion_r697554603

File: pkg/envoy/rds.go [hunk 3]
```diff
@@ -53,26 +54,26 @@ func (l RouteDomainCheck) Run() outcomes.Outcome {
 		for _, virtualHost := range dynRouteCfg.GetVirtualHosts() {
 			for _, domain := range virtualHost.GetDomains() {
 				foundAnyRouteDomains = true
-				if l.Domain == "" {
+				if len(l.Domains) == 0 {
```
Comment (SanyaKochhar) [inline pkg/envoy/rds.go:57]: could `l` be renamed to `check` for clarity? A lot of the other checks use this and `l` is a bit ambiguous

File: pkg/envoy/rds.go [hunk 4]
```diff
@@ -94,21 +95,35 @@ func (l RouteDomainCheck) Description() string {
...
+// whether the Envoy config has inbound dynamic route domains from the Pod's services.
+func HasInboundDynamicRouteConfigDomainPodCheck(client kubernetes.Interface, configGetter ConfigGetter, pod *corev1.Pod) RouteDomainCheck {
+	return ArePodServicesInDynamicRouteConfigDomains(client, configGetter, pod, InboundDynamicRouteConfigName)
 }
 
-// NewInboundRouteDomainPodCheck creates a DestinationEndpointCheck which checks whether the Envoy config has an inbound dynamic route domain to the Pod
-func NewInboundRouteDomainPodCheck(configGetter ConfigGetter, pod *corev1.Pod) RouteDomainCheck {
+// ArePodServicesInDynamicRouteConfigDomains checks whether the pod's corresponding service's domains are
+// contained in the envoy dynamic route config domain list.
+func ArePodServicesInDynamicRouteConfigDomains(client kubernetes.Interface, configGetter ConfigGetter, pod *corev1.Pod, routeName string) RouteDomainCheck {
```
Comment (SanyaKochhar) [inline pkg/envoy/rds.go:112]: could this name be simplified? `ArePodServicesInDynamicRouteConfigDomains`

**Review by SanyaKochhar — CHANGES_REQUESTED** [review]

File: pkg/envoy/rds.go (hunk 3 above)
Comment (johnsonshi) [reply pkg/envoy/rds.go:57]: Sure! Thanks!

File: pkg/envoy/rds.go (hunk 4 above)
Comment (johnsonshi) [reply pkg/envoy/rds.go:112]: Hmmm I think we should leave this for now - I can't figure out a better name.

Comment (SanyaKochhar) [reply pkg/envoy/rds.go:112]: How about `NewPodServicesRouteDomainCheck`  to parallel with `NewInboundRouteDomainPodCheck`

Comment (SanyaKochhar) [reply pkg/envoy/rds.go:112]: or rather NewServicesRouteDomainPodCheck -- not sure which follows more logically

**Review by SanyaKochhar — APPROVED** [review]

File: pkg/envoy/rds.go (hunk 4 above)
Comment (johnsonshi) [reply pkg/envoy/rds.go]: Done! Chose `NewPodServicesRouteDomainCheck`

**Comment by shashankram** [comment]
@johnsonshi could you describe in the commit and PR description what's being fixed (as noted in https://github.com/openservicemesh/osm-health/pull/71#issuecomment-903689754).

File: pkg/envoy/rds.go [hunk 5]
```diff
@@ -46,69 +47,83 @@ func (l RouteDomainCheck) Run() outcomes.Outcome {
...
 					break
 				}
-				if domain == l.Domain {
+				if _, ok := check.Domains[domain]; ok {
 					foundSpecificRouteDomain = true
 					break
 				}
 			}
-			if (l.Domain == "" && foundAnyRouteDomains) || foundSpecificRouteDomain {
+			if (len(check.Domains) == 0 && foundAnyRouteDomains) || foundSpecificRouteDomain {
```
Comment (jaellio) [inline pkg/envoy/rds.go:65]: nit: In the OSM repo we primarily compare strings with an empty string. For consistency, we should continue to use the empty string comparison.

**Review by jaellio — APPROVED** [review]

**Comment by johnsonshi** [comment]
@shashankram added the PR description. RDS checks now succeed as well:

image

**Pull request merged**